osu.Game/Database/StandardisedScoreMigrationTools.cs

```diff
@@ -0,0 +1,178 @@
+                                         .ToList();
+
+            // This is a list of all results, ordered from best to worst.
+            // We are constructing a "best possible" score from the statistics provided because it's the best we can do.
+            List<HitResult> sortedHits = score.Statistics
+                                              .Where(kvp => kvp.Key.AffectsCombo())
+                                              .OrderByDescending(kvp => Judgement.ToNumericResult(kvp.Key))
+                                              .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value))
+                                              .ToList();
+
+            if (maximumJudgements.Count != sortedHits.Count)
+            {
+                // Older scores may not have maximum judgements populated correctly.
+                // In this case we need to fill them.
+                maximumJudgements = sortedHits
+                                    .Select(r => new FakeJudgement(getMaxJudgementFor(r, maxRulesetJudgement)))
+                                    .ToList();
+            }
+
+            foreach (var judgement in maximumJudgements)
+                beatmap.HitObjects.Add(new FakeHit(judgement));
+
+            processor.ApplyBeatmap(beatmap);
+
+            Queue<HitResult> misses = new Queue<HitResult>(score.Statistics
+                                                                .Where(kvp => kvp.Key == HitResult.Miss || kvp.Key == HitResult.LargeTickMiss)
+                                                                .SelectMany(kvp => Enumerable.Repeat(kvp.Key, kvp.Value)));
+
+            int maxJudgementIndex = 0;
+
+            foreach (var result in sortedHits)
+            {
+                // misses are handled from the queue.
+                if (result == HitResult.Miss || result == HitResult.LargeTickMiss)
+                    continue;
+
+                if (processor.Combo.Value == maxCombo)
+                {
+                    if (misses.Count > 0)
+                    {
+                        processor.ApplyResult(new JudgementResult(null!, maximumJudgements[maxJudgementIndex++])
+                        {
+                            Type = misses.Dequeue(),
+                        });
+                    }
+                    else
+                    {
+                        // worst case scenario, insert a miss.
+                        processor.ApplyResult(new JudgementResult(null!, new FakeJudgement(getMaxJudgementFor(HitResult.Miss, maxRulesetJudgement)))
+                        {
+                            Type = HitResult.Miss,
+                        });
+                    }
+                }
+
+                processor.ApplyResult(new JudgementResult(null!, maximumJudgements[maxJudgementIndex++])
+                {
+                    Type = result
+                });
```

Comment:
This runs unconditionally after the `processor.Combo.Value == maxCombo` condition. This means that `maxJudgementIndex` can be incremented twice in a single loop of this method. Is this expected?

Reply:
Yep. This is the actual hit. The previous one is if we need to insert a miss.

To make it clear: we have no idea when the misses happened, so we play through the score and insert to ensure the user never exceeds the databased `MaxCombo`.